File: ios/Classes/NearbyCrossPlugin.swift
import Flutter
import UIKit


public class NearbyCrossPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel
    private var context: UIApplication
    private var callbacks: NearbyCrossCallbacks
    private var advertiser: NCAdvertiser?
    private var discoverer: NCDiscoverer?

    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
        self.context = UIApplication.shared
        self.callbacks = NearbyCrossCallbacks(channel: channel)
        super.init()
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: GeneralConstants.PLUGIN_NAME, binaryMessenger: registrar.messenger())
        let instance = NearbyCrossPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case ChannelMethods.GET_PLATFORM_VERSION:
          result("iOS " + UIDevice.current.systemVersion)
        case ChannelMethods.START_ADVERTISING:
            guard let args = call.arguments as? [String: Any],
                  let serviceId = args["serviceId"] as? String,
                  let userName = args["username"] as? FlutterStandardTypedData,
                  let strategy = args["strategy"] as? String,
                  let manualAcceptConnections = args["manualAcceptConnections"] as? String  else {
                      result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                      return
            }
            if (advertiser == nil) {
                advertiser = NCAdvertiser(serviceId: serviceId, strategy: strategy, context: context, callbacks: callbacks.advertiser, userName: userName, manualAcceptConnections:  manualAcceptConnections == "1")
            }
            
            advertiser?.startAdvertising()
            result(nil)
        case ChannelMethods.START_DISCOVERY:
            guard let args = call.arguments as? [String: Any],
                  let serviceId = args["serviceId"] as? String,
                  let userName = args["username"] as? FlutterStandardTypedData,
                  let strategy = args["strategy"] as? String else {
                      result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                      return
            }
            if (discoverer == nil) {
                discoverer = NCDiscoverer(serviceId: serviceId, strategy: strategy, context: context, callbacks: callbacks.discoverer, userName: userName)
            }
            
            discoverer?.startDiscovering()
            result(nil)
        case ChannelMethods.CONNECT:
            guard let args = call.arguments as? [String: Any],
                  let endpointId = args["endpointId"] as? String else {
                result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                return
            }
            
            discoverer?.connect(endpointId: endpointId)
            result(nil)
        case ChannelMethods.SEND_DATA:
            guard let args = call.arguments as? [String: Any],
                  let data = args["data"] as? FlutterStandardTypedData,
                  let endpointId = args["endpointId"] as? String else {
                result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                return
            }
            
            if (discoverer != nil) {
                var _ = discoverer?.sendData(data: Data(data.data), to: endpointId)
            } else if (advertiser != nil) {
                var _ = advertiser?.sendData(data: Data(data.data), to: endpointId)
            }

            result(nil)
        case ChannelMethods.ACCEPT_CONNECTION:
            guard let args = call.arguments as? [String: Any],
                  let endpointId = args["endpointId"] as? String else {
                result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                return
            }
            
            advertiser?.acceptConnection(endpointId: endpointId)
            result(nil)
            break;
        case ChannelMethods.REJECT_CONNECTION:
            guard let args = call.arguments as? [String: Any],
                  let endpointId = args["endpointId"] as? String else {
                result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                return
            }
            
            advertiser?.rejectConnection(endpointId: endpointId)
            result(nil)
            break;
        case ChannelMethods.DISCONNECT_FROM:
            guard let args = call.arguments as? [String: Any],
                  let endpointId = args["endpointId"] as? String else {
                result(FlutterError(code: "argument_error", message: "Missing arguments", details: nil))
                return
            }
            
            advertiser?.disconnectFrom(from: endpointId)
            discoverer?.disconnectFrom(from: endpointId)
            result(nil)
        case ChannelMethods.STOP_DISCOVERING:
            discoverer?.stopDiscovering()
            result(nil)
        case ChannelMethods.STOP_ADVERTISING:
            advertiser?.stopAdvertising()
            result(nil)
        case ChannelMethods.STOP_ALL_CONNECTIONS:
           advertiser?.stopAllConnections()
           discoverer?.stopAllConnections()
           result(nil)
        default:
          result("Not implemented")
        }
        
    }
}

File: android/src/main/kotlin/com/example/nearby_cross/NearbyCrossPlugin.kt
package com.example.nearby_cross

import android.content.Context
import android.os.Build
import androidx.annotation.NonNull
import com.example.nearby_cross.callbacks.NearbyCrossCallbacks
import com.example.nearby_cross.constants.ChannelMethods
import com.example.nearby_cross.constants.Constants
import com.example.nearby_cross.utils.InterfaceUtils
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** NearbyCrossPlugin */
class NearbyCrossPlugin : FlutterPlugin, MethodCallHandler {

    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var callbacks: NearbyCrossCallbacks
    private var advertiser: Advertiser? = null
    private var discoverer: Discoverer? = null
    private var dataManager: DataManager = DataManager()

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, Constants.PLUGIN_NAME)
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        callbacks = NearbyCrossCallbacks(channel)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            ChannelMethods.GET_PLATFORM_VERSION -> {
                result.success("Android ${Build.VERSION.RELEASE}")
            }
            ChannelMethods.START_DISCOVERY -> {
                val serviceId = call.argument<String>("serviceId")
                val userName = call.argument<ByteArray>("username") as ByteArray
                val strategy = call.argument<String>("strategy")
                this.discoverer = Discoverer(
                    serviceId as String,
                    strategy as String,
                    context,
                    callbacks.discoverer,
                    userName,
                )
                this.discoverer?.startDiscovering(context)
                result.success(null)
            }
            ChannelMethods.START_ADVERTISING -> {
                val serviceId = call.argument<String>("serviceId")
                val userName = call.argument<ByteArray>("username") as ByteArray
                val strategy = call.argument<String>("strategy")
                val manualAcceptConnections =
                    InterfaceUtils.parseStringAsBoolean(call.argument<String>("manualAcceptConnections") as String)
                this.advertiser = Advertiser(
                    serviceId as String,
                    strategy as String,
                    context,
                    callbacks.advertiser,
                    userName,
                    manualAcceptConnections
                )
                this.advertiser?.startAdvertising(context)
                result.success(null)
            }
            ChannelMethods.STOP_DISCOVERING -> {
                this.discoverer?.stopDiscovering(context)
                result.success(null)
            }
            ChannelMethods.STOP_ADVERTISING -> {
                this.advertiser?.stopAdvertising(context)
                result.success(null)
            }
            ChannelMethods.CONNECT -> {
                val endpointId = call.argument<String>("endpointId")
                this.discoverer?.connect(endpointId as String)
            }
            ChannelMethods.STOP_ALL_CONNECTIONS -> {
                this.advertiser?.stopAllConnections(context)
                this.discoverer?.stopAllConnections(context)
                result.success(null)
            }
            ChannelMethods.DISCONNECT_FROM -> {
                val endpointId = call.argument<String>("endpointId")
                this.advertiser?.disconnectFromEndpointId(context, endpointId as String)
                this.discoverer?.disconnectFromEndpointId(context, endpointId as String)
                result.success(null)
            }
            ChannelMethods.SEND_DATA -> {
                val data = call.argument<ByteArray>("data") as ByteArray
                val endpointId = call.argument<String>("endpointId") as String
                this.dataManager.sendData(context, data, endpointId)
                result.success(null)
            }
            ChannelMethods.ACCEPT_CONNECTION -> {
                val endpointId = call.argument<String>("endpointId") as String
                this.advertiser?.acceptConnection(endpointId)
                result.success(null)
            }
            ChannelMethods.REJECT_CONNECTION -> {
                val endpointId = call.argument<String>("endpointId") as String
                this.advertiser?.rejectConnection(endpointId)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}


